In [52]:
from functools import reduce
from operator import add

import numpy as np
import pandas as pd

# MRIO

## Z: Inter-industry transactions of intermediate products (unit: €/year)

Intermediate products: goods and services as inputs in the production processes.

In [53]:
# Z (unit: Million euro)
country = 'Netherlands'

regions = [country, 'OECD', 'NONOECD']
sectors = ['Agriculture', 'Industry', 'EHP', 'RoService']

Z_columns = pd.MultiIndex.from_product([regions, sectors])


Z_values_NL = [
    [1053,16373,170,787,639,2112,131,290,318,963,75,247],
    [3745,88955,11478,37202,1474,53747,5006,11973,1806,24852,2795,8394],
    [194,2757,6868,7736,8,622,424,926,47,408,101,250],
    [6580,83484,30315,194654,543,32537,8653,51600,562,7212,2017,10622],
    [332,2256,195,280,91501,644731,12965,45731,7680,73150,2058,8743],
    [1494,54739,3588,10189,147986,8557780,1134269,1956688,36416,1587100,81398,236817],
    [5,50,41,127,3822,115135,444681,206710,99,2376,407,767],
    [1066,13323,5314,27692,445915,5900875,3094832,10924341,10928,161156,33899,183560],
    [1555,4626,308,655,9846,66276,3381,8173,491691,1816071,43368,353492],
    [1386,45857,3241,6197,15006,1326856,140554,188453,554841,18884285,1146491,2834820],
    [3,35,18,49,41,1127,2189,1380,4201,54153,88652,58534],
    [232,3264,929,6807,3650,97082,65691,209966,428179,5603261,1309967,6031464],
]

Z_values = Z_values_NL
Z = pd.DataFrame(Z_values, index=Z_columns, columns=Z_columns)
Z

Netherlands                                  OECD  \
                        Agriculture Industry    EHP RoService Agriculture   
Netherlands Agriculture        1053    16373    170       787         639   
            Industry           3745    88955  11478     37202        1474   
            EHP                 194     2757   6868      7736           8   
            RoService          6580    83484  30315    194654         543   
OECD        Agriculture         332     2256    195       280       91501   
            Industry           1494    54739   3588     10189      147986   
            EHP                   5       50     41       127        3822   
            RoService          1066    13323   5314     27692      445915   
NONOECD     Agriculture        1555     4626    308       655        9846   
            Industry           1386    45857   3241      6197       15006   
            EHP                   3       35     18        49          41   
            RoService           232     3264    929      6807        3650   

                                                        NONOECD            \
                        Industry      EHP RoService Agriculture  Industry   
Netherlands Agriculture     2112      131       290         318       963   
            Industry       53747     5006     11973        1806     24852   
            EHP              622      424       926          47       408   
            RoService      32537     8653     51600         562      7212   
OECD        Agriculture   644731    12965     45731        7680     73150   
            Industry     8557780  1134269   1956688       36416   1587100   
            EHP           115135   444681    206710          99      2376   
            RoService    5900875  3094832  10924341       10928    161156   
NONOECD     Agriculture    66276     3381      8173      491691   1816071   
            Industry     1326856   140554    188453      554841  18884285   
            EHP             1127     2189      1380        4201     54153   
            RoService      97082    65691    209966      428179   5603261   

                                            
                             EHP RoService  
Netherlands Agriculture       75       247  
            Industry        2795      8394  
            EHP              101       250  
            RoService       2017     10622  
OECD        Agriculture     2058      8743  
            Industry       81398    236817  
            EHP              407       767  
            RoService      33899    183560  
NONOECD     Agriculture    43368    353492  
            Industry     1146491   2834820  
            EHP            88652     58534  
            RoService    1309967   6031464

Q1: What does the value in cell C8 (3,0) tell you? (2 points)  

In year 2019, Dutch Agriculture purchased 6580 milliion euros worth of RoService products for production.  
//  
The monetary transaction flow (in Million euro) supplied by the (domestic) "Rest of Services" sector and purchased by the (domestic) Agriculture sector 

## Y: Final demand (unit: €/year)

In [54]:
# Y (unit: Million euro)
Y_columns = regions

Y_values_NL = [
    [2257,6661,1352],
    [127940,75538,39231],
    [176644,4150,1311],
    [291297,40362,7982],
    [2938,328169,32997],
    [57759,11268697,1319292],
    [875,11903608,31982],
    [15131,16922388,198169],
    [6130,89325,1376213],
    [39556,1406289,14122275],
    [675,42575,7547773],
    [4525,281639,10337291],
]

Y_values = Y_values_NL
Y = pd.DataFrame(Y_values, index=Z_columns, columns=Y_columns)
Y

Netherlands      OECD   NONOECD
Netherlands Agriculture         2257      6661      1352
            Industry          127940     75538     39231
            EHP               176644      4150      1311
            RoService         291297     40362      7982
OECD        Agriculture         2938    328169     32997
            Industry           57759  11268697   1319292
            EHP                  875  11903608     31982
            RoService          15131  16922388    198169
NONOECD     Agriculture         6130     89325   1376213
            Industry           39556   1406289  14122275
            EHP                  675     42575   7547773
            RoService           4525    281639  10337291

## x: Gross output (unit: €/year)

Gross output of Sector j: intermediate products + finished products of Sector j

In [55]:
# x (unit: Million euro)
x = Z.sum(axis=1) + Y.sum(axis=1)
x

Netherlands  Agriculture       33428
             Industry         494136
             EHP              202446
             RoService        768420
OECD         Agriculture     1253726
             Industry       26454212
             EHP            12710685
             RoService      37938589
NONOECD      Agriculture     4271110
             Industry       40716107
             EHP             7801405
             RoService      24383947
dtype: int64

## V: Value added (unit: €/year)

- Economic gains generated by producing goods and services.
- Main components: wages, taxes minus subsidies, capital profits, etc
- v: value added pet unit per year
- V: value added per year

In [56]:
# # v (Unit: Million euro/Million euro)
# v_values_NL = []  # len = Z_columns

# v_values = v_values_NL

# v = pd.Series(v_values, index=Z_columns)
# v

In [57]:
# # V (Unit: Million euro)
# # method 1
# V = x.mul(v)
# V

In [58]:
# !!! method 2: required to have calculated M import
# See below
# V2 = x - Z.sum(axis=1) - M_INT.sum(axis=1)
# V2

<hr>

## A: Technical coefficient (unit: €/€)

$$a_{i,j} = z_{i,j}/x_j$$
$$z_{i,j} = a_{i,j} x_j$$

$aij$ : direct requirements of sector i per € output of sector j  
$lij$ : total requirements of sector i per € final demand of sector j

In [43]:
# A (Unit: Million euro/Million euro)
A = Z.divide(x)  # Z = A.mul(x)
A.round(4)

Netherlands                                   OECD  \
                        Agriculture Industry     EHP RoService Agriculture   
Netherlands Agriculture      0.0315   0.0331  0.0008    0.0010      0.0005   
            Industry         0.1120   0.1800  0.0567    0.0484      0.0012   
            EHP              0.0058   0.0056  0.0339    0.0101      0.0000   
            RoService        0.1968   0.1689  0.1497    0.2533      0.0004   
OECD        Agriculture      0.0099   0.0046  0.0010    0.0004      0.0730   
            Industry         0.0447   0.1108  0.0177    0.0133      0.1180   
            EHP              0.0001   0.0001  0.0002    0.0002      0.0030   
            RoService        0.0319   0.0270  0.0262    0.0360      0.3557   
NONOECD     Agriculture      0.0465   0.0094  0.0015    0.0009      0.0079   
            Industry         0.0415   0.0928  0.0160    0.0081      0.0120   
            EHP              0.0001   0.0001  0.0001    0.0001      0.0000   
            RoService        0.0069   0.0066  0.0046    0.0089      0.0029   

                                                       NONOECD           \
                        Industry     EHP RoService Agriculture Industry   
Netherlands Agriculture   0.0001  0.0000    0.0000      0.0001   0.0000   
            Industry      0.0020  0.0004    0.0003      0.0004   0.0006   
            EHP           0.0000  0.0000    0.0000      0.0000   0.0000   
            RoService     0.0012  0.0007    0.0014      0.0001   0.0002   
OECD        Agriculture   0.0244  0.0010    0.0012      0.0018   0.0018   
            Industry      0.3235  0.0892    0.0516      0.0085   0.0390   
            EHP           0.0044  0.0350    0.0054      0.0000   0.0001   
            RoService     0.2231  0.2435    0.2879      0.0026   0.0040   
NONOECD     Agriculture   0.0025  0.0003    0.0002      0.1151   0.0446   
            Industry      0.0502  0.0111    0.0050      0.1299   0.4638   
            EHP           0.0000  0.0002    0.0000      0.0010   0.0013   
            RoService     0.0037  0.0052    0.0055      0.1003   0.1376   

                                           
                            EHP RoService  
Netherlands Agriculture  0.0000    0.0000  
            Industry     0.0004    0.0003  
            EHP          0.0000    0.0000  
            RoService    0.0003    0.0004  
OECD        Agriculture  0.0003    0.0004  
            Industry     0.0104    0.0097  
            EHP          0.0001    0.0000  
            RoService    0.0043    0.0075  
NONOECD     Agriculture  0.0056    0.0145  
            Industry     0.1470    0.1163  
            EHP          0.0114    0.0024  
            RoService    0.1679    0.2474

In [61]:
A.sum()

Netherlands  Agriculture    0.527851
             Industry       0.638931
             EHP            0.308551
             RoService      0.380489
OECD         Agriculture    0.574632
             Industry       0.635017
             EHP            0.386508
             RoService      0.358638
NONOECD      Agriculture    0.359805
             Industry       0.692969
             EHP            0.347531
             RoService      0.398939
dtype: float64

In [44]:
Id = pd.DataFrame(np.identity(len(Z_columns)), index=A.index, columns=A.columns)
Id

Netherlands                                OECD  \
                        Agriculture Industry  EHP RoService Agriculture   
Netherlands Agriculture         1.0      0.0  0.0       0.0         0.0   
            Industry            0.0      1.0  0.0       0.0         0.0   
            EHP                 0.0      0.0  1.0       0.0         0.0   
            RoService           0.0      0.0  0.0       1.0         0.0   
OECD        Agriculture         0.0      0.0  0.0       0.0         1.0   
            Industry            0.0      0.0  0.0       0.0         0.0   
            EHP                 0.0      0.0  0.0       0.0         0.0   
            RoService           0.0      0.0  0.0       0.0         0.0   
NONOECD     Agriculture         0.0      0.0  0.0       0.0         0.0   
            Industry            0.0      0.0  0.0       0.0         0.0   
            EHP                 0.0      0.0  0.0       0.0         0.0   
            RoService           0.0      0.0  0.0       0.0         0.0   

                                                    NONOECD                \
                        Industry  EHP RoService Agriculture Industry  EHP   
Netherlands Agriculture      0.0  0.0       0.0         0.0      0.0  0.0   
            Industry         0.0  0.0       0.0         0.0      0.0  0.0   
            EHP              0.0  0.0       0.0         0.0      0.0  0.0   
            RoService        0.0  0.0       0.0         0.0      0.0  0.0   
OECD        Agriculture      0.0  0.0       0.0         0.0      0.0  0.0   
            Industry         1.0  0.0       0.0         0.0      0.0  0.0   
            EHP              0.0  1.0       0.0         0.0      0.0  0.0   
            RoService        0.0  0.0       1.0         0.0      0.0  0.0   
NONOECD     Agriculture      0.0  0.0       0.0         1.0      0.0  0.0   
            Industry         0.0  0.0       0.0         0.0      1.0  0.0   
            EHP              0.0  0.0       0.0         0.0      0.0  1.0   
            RoService        0.0  0.0       0.0         0.0      0.0  0.0   

                                   
                        RoService  
Netherlands Agriculture       0.0  
            Industry          0.0  
            EHP               0.0  
            RoService         0.0  
OECD        Agriculture       0.0  
            Industry          0.0  
            EHP               0.0  
            RoService         0.0  
NONOECD     Agriculture       0.0  
            Industry          0.0  
            EHP               0.0  
            RoService         1.0

## L: Leontief inverse
### by inverse

$Z + y = x$  
$Ax + y = x$  
$y = (I-A)x$  
$(I-A)^{-1}(I-A)x=(I-A)^{-1}y$  

$x = (I-A)^{-1}y = Ly$

$lij$ : total requirements of sector i per € final demand of sector j

In [45]:
(Id - A)

Netherlands                                      OECD  \
                        Agriculture  Industry       EHP RoService Agriculture   
Netherlands Agriculture    0.968499 -0.033135 -0.000840 -0.001024   -0.000510   
            Industry      -0.112032  0.819979 -0.056697 -0.048414   -0.001176   
            EHP           -0.005804 -0.005579  0.966075 -0.010067   -0.000006   
            RoService     -0.196841 -0.168949 -0.149744  0.746683   -0.000433   
OECD        Agriculture   -0.009932 -0.004566 -0.000963 -0.000364    0.927017   
            Industry      -0.044693 -0.110777 -0.017723 -0.013260   -0.118037   
            EHP           -0.000150 -0.000101 -0.000203 -0.000165   -0.003049   
            RoService     -0.031889 -0.026962 -0.026249 -0.036038   -0.355672   
NONOECD     Agriculture   -0.046518 -0.009362 -0.001521 -0.000852   -0.007853   
            Industry      -0.041462 -0.092802 -0.016009 -0.008065   -0.011969   
            EHP           -0.000090 -0.000071 -0.000089 -0.000064   -0.000033   
            RoService     -0.006940 -0.006605 -0.004589 -0.008858   -0.002911   

                                                          NONOECD            \
                         Industry       EHP RoService Agriculture  Industry   
Netherlands Agriculture -0.000080 -0.000010 -0.000008   -0.000074 -0.000024   
            Industry    -0.002032 -0.000394 -0.000316   -0.000423 -0.000610   
            EHP         -0.000024 -0.000033 -0.000024   -0.000011 -0.000010   
            RoService   -0.001230 -0.000681 -0.001360   -0.000132 -0.000177   
OECD        Agriculture -0.024372 -0.001020 -0.001205   -0.001798 -0.001797   
            Industry     0.676506 -0.089237 -0.051575   -0.008526 -0.038980   
            EHP         -0.004352  0.965015 -0.005449   -0.000023 -0.000058   
            RoService   -0.223060 -0.243483  0.712052   -0.002559 -0.003958   
NONOECD     Agriculture -0.002505 -0.000266 -0.000215    0.884880 -0.044603   
            Industry    -0.050157 -0.011058 -0.004967   -0.129906  0.536196   
            EHP         -0.000043 -0.000172 -0.000036   -0.000984 -0.001330   
            RoService   -0.003670 -0.005168 -0.005534   -0.100250 -0.137618   

                                             
                              EHP RoService  
Netherlands Agriculture -0.000010 -0.000010  
            Industry    -0.000358 -0.000344  
            EHP         -0.000013 -0.000010  
            RoService   -0.000259 -0.000436  
OECD        Agriculture -0.000264 -0.000359  
            Industry    -0.010434 -0.009712  
            EHP         -0.000052 -0.000031  
            RoService   -0.004345 -0.007528  
NONOECD     Agriculture -0.005559 -0.014497  
            Industry    -0.146960 -0.116258  
            EHP          0.988636 -0.002401  
            RoService   -0.167914  0.752646

In [46]:
L = pd.DataFrame(np.linalg.inv((Id - A).values), A.columns, A.index)
L.round(3)

Netherlands                                  OECD  \
                        Agriculture Industry    EHP RoService Agriculture   
Netherlands Agriculture       1.038    0.043  0.004     0.004       0.001   
            Industry          0.162    1.245  0.086     0.082       0.003   
            EHP               0.010    0.011  1.038     0.015       0.000   
            RoService         0.313    0.296  0.229     1.362       0.003   
OECD        Agriculture       0.016    0.014  0.003     0.003       1.086   
            Industry          0.124    0.244  0.056     0.052       0.243   
            EHP               0.002    0.002  0.001     0.001       0.008   
            RoService         0.118    0.151  0.074     0.091       0.622   
NONOECD     Agriculture       0.066    0.031  0.007     0.006       0.014   
            Industry          0.154    0.271  0.062     0.049       0.062   
            EHP               0.001    0.001  0.000     0.000       0.000   
            RoService         0.053    0.071  0.023     0.028       0.023   

                                                      NONOECD                  \
                        Industry    EHP RoService Agriculture Industry    EHP   
Netherlands Agriculture    0.000  0.000     0.000       0.000    0.000  0.000   
            Industry       0.005  0.001     0.001       0.001    0.002  0.001   
            EHP            0.000  0.000     0.000       0.000    0.000  0.000   
            RoService      0.005  0.002     0.003       0.001    0.002  0.001   
OECD        Agriculture    0.041  0.006     0.005       0.004    0.008  0.002   
            Industry       1.536  0.173     0.114       0.039    0.128  0.043   
            EHP            0.010  1.039     0.009       0.000    0.001  0.000   
            RoService      0.507  0.413     1.446       0.023    0.060  0.026   
NONOECD     Agriculture    0.014  0.004     0.002       1.150    0.107  0.029   
            Industry       0.163  0.047     0.029       0.330    1.987  0.353   
            EHP            0.000  0.000     0.000       0.002    0.004  1.013   
            RoService      0.043  0.020     0.017       0.214    0.380  0.295   

                                   
                        RoService  
Netherlands Agriculture     0.000  
            Industry        0.001  
            EHP             0.000  
            RoService       0.001  
OECD        Agriculture     0.002  
            Industry        0.042  
            EHP             0.000  
            RoService       0.031  
NONOECD     Agriculture     0.039  
            Industry        0.317  
            EHP             0.004  
            RoService       1.393

In [64]:
L.sum()

Netherlands  Agriculture    2.056194
             Industry       2.379542
             EHP            1.584098
             RoService      1.691763
OECD         Agriculture    2.065139
             Industry       2.325774
             EHP            1.707630
             RoService      1.627316
NONOECD      Agriculture    1.765600
             Industry       2.677850
             EHP            1.764214
             RoService      1.831374
dtype: float64

### by series expansion

L covers supply chains of inifinte lengths. Power expansions stops at a certain depth

$$(I-A)^{-1} = I + A + A^2 + A^3 + ... + A^n + ... = \lim_{n \to +\infty} \sum_{k=0}^{n} A^k$$

The power series converges only under certain conditions:
- A ≥ 0 (coefficients matrix contains only non-negative terms)
- The system produces more output than it requires inputs N(A) < 1
- |I − A| > 0
- More generally, the [Hawkins-Simon conditions](https://www.wikiwand.com/en/Hawkins%E2%80%93Simon_condition)

the limit sum convergs if all of the eigenvalues of A have absolute value smaller than 1 (?)

In [12]:
# AA
A.dot(A).round(4)

Netherlands                                   OECD  \
                        Agriculture Industry     EHP RoService Agriculture   
Netherlands Agriculture      0.0049   0.0072  0.0021    0.0019      0.0001   
            Industry         0.0337   0.0449  0.0195    0.0217      0.0007   
            EHP              0.0030   0.0031  0.0030    0.0032      0.0000   
            RoService        0.0760   0.0808  0.0528    0.0741      0.0011   
OECD        Agriculture      0.0029   0.0045  0.0009    0.0007      0.0087   
            Industry         0.0359   0.0653  0.0168    0.0156      0.0661   
            EHP              0.0005   0.0007  0.0003    0.0003      0.0028   
            RoService        0.0343   0.0467  0.0198    0.0243      0.1556   
NONOECD     Agriculture      0.0102   0.0090  0.0018    0.0014      0.0025   
            Industry         0.0420   0.0703  0.0162    0.0125      0.0156   
            EHP              0.0001   0.0002  0.0001    0.0001      0.0001   
            RoService        0.0152   0.0189  0.0056    0.0063      0.0058   

                                                       NONOECD           \
                        Industry     EHP RoService Agriculture Industry   
Netherlands Agriculture   0.0001  0.0000    0.0000      0.0000   0.0000   
            Industry      0.0012  0.0004    0.0003      0.0003   0.0006   
            EHP           0.0000  0.0000    0.0000      0.0000   0.0000   
            RoService     0.0014  0.0007    0.0009      0.0002   0.0004   
OECD        Agriculture   0.0100  0.0026    0.0017      0.0008   0.0021   
            Industry      0.1217  0.0452    0.0325      0.0102   0.0329   
            EHP           0.0029  0.0029    0.0020      0.0001   0.0002   
            RoService     0.1465  0.0990    0.0963      0.0049   0.0135   
NONOECD     Agriculture   0.0037  0.0009    0.0005      0.0205   0.0279   
            Industry      0.0419  0.0119    0.0071      0.0875   0.2392   
            EHP           0.0001  0.0000    0.0000      0.0005   0.0010   
            RoService     0.0106  0.0047    0.0039      0.0544   0.1027   

                                           
                            EHP RoService  
Netherlands Agriculture  0.0000    0.0000  
            Industry     0.0003    0.0003  
            EHP          0.0000    0.0000  
            RoService    0.0003    0.0003  
OECD        Agriculture  0.0006    0.0006  
            Industry     0.0112    0.0107  
            EHP          0.0001    0.0001  
            RoService    0.0056    0.0069  
NONOECD     Agriculture  0.0097    0.0105  
            Industry     0.0907    0.0855  
            EHP          0.0007    0.0008  
            RoService    0.0643    0.0791

In [13]:
# AAA
A.dot(A.dot(A)).round(4)

Netherlands                                   OECD  \
                        Agriculture Industry     EHP RoService Agriculture   
Netherlands Agriculture      0.0014   0.0018  0.0008    0.0009      0.0000   
            Industry         0.0106   0.0132  0.0065    0.0079      0.0004   
            EHP              0.0011   0.0012  0.0008    0.0010      0.0000   
            RoService        0.0265   0.0301  0.0176    0.0234      0.0007   
OECD        Agriculture      0.0015   0.0024  0.0007    0.0006      0.0025   
            Industry         0.0207   0.0335  0.0103    0.0105      0.0315   
            EHP              0.0004   0.0006  0.0002    0.0002      0.0013   
            RoService        0.0232   0.0347  0.0125    0.0143      0.0635   
NONOECD     Agriculture      0.0040   0.0055  0.0014    0.0012      0.0013   
            Industry         0.0286   0.0447  0.0118    0.0104      0.0126   
            EHP              0.0001   0.0002  0.0000    0.0000      0.0000   
            RoService        0.0119   0.0169  0.0046    0.0044      0.0050   

                                                       NONOECD           \
                        Industry     EHP RoService Agriculture Industry   
Netherlands Agriculture   0.0001  0.0000    0.0000      0.0000   0.0000   
            Industry      0.0006  0.0002    0.0002      0.0002   0.0004   
            EHP           0.0000  0.0000    0.0000      0.0000   0.0000   
            RoService     0.0010  0.0005    0.0005      0.0002   0.0003   
OECD        Agriculture   0.0040  0.0014    0.0011      0.0005   0.0015   
            Industry      0.0503  0.0209    0.0162      0.0078   0.0222   
            EHP           0.0015  0.0008    0.0007      0.0001   0.0002   
            RoService     0.0739  0.0404    0.0362      0.0048   0.0138   
NONOECD     Agriculture   0.0029  0.0009    0.0006      0.0071   0.0155   
            Industry      0.0283  0.0091    0.0060      0.0502   0.1284   
            EHP           0.0001  0.0000    0.0000      0.0003   0.0006   
            RoService     0.0101  0.0037    0.0027      0.0277   0.0615   

                                           
                            EHP RoService  
Netherlands Agriculture  0.0000    0.0000  
            Industry     0.0002    0.0002  
            EHP          0.0000    0.0000  
            RoService    0.0002    0.0002  
OECD        Agriculture  0.0005    0.0005  
            Industry     0.0083    0.0081  
            EHP          0.0001    0.0001  
            RoService    0.0052    0.0056  
NONOECD     Agriculture  0.0061    0.0062  
            Industry     0.0515    0.0509  
            EHP          0.0003    0.0003  
            RoService    0.0296    0.0326

In [14]:
def leontief_power_series(A_mat, depth):
    Id_mat = pd.DataFrame(np.identity(len(Z_columns)), index=A.index, columns=A.columns)
    
    res = [None]*(depth+1)
    res[0] = Id_mat

    for i in range(1, depth+1):
        res[i] = res[i-1].dot(A_mat)
        # print(f"{'A'*i}\n{res[i]}\n\n")

    return reduce(add, res)

In [15]:
leontief_power_series(A, 3).round(4)

Netherlands                                   OECD  \
                        Agriculture Industry     EHP RoService Agriculture   
Netherlands Agriculture      1.0378   0.0421  0.0037    0.0038      0.0007   
            Industry         0.1563   1.2381  0.0828    0.0781      0.0023   
            EHP              0.0099   0.0099  1.0377    0.0142      0.0001   
            RoService        0.2993   0.2798  0.2202    1.3508      0.0022   
OECD        Agriculture      0.0143   0.0115  0.0026    0.0017      1.0841   
            Industry         0.1013   0.2096  0.0448    0.0393      0.2157   
            EHP              0.0010   0.0014  0.0007    0.0007      0.0071   
            RoService        0.0894   0.1083  0.0586    0.0747      0.5748   
NONOECD     Agriculture      0.0607   0.0239  0.0047    0.0034      0.0117   
            Industry         0.1121   0.2079  0.0440    0.0309      0.0402   
            EHP              0.0004   0.0004  0.0002    0.0002      0.0001   
            RoService        0.0340   0.0423  0.0148    0.0196      0.0137   

                                                       NONOECD           \
                        Industry     EHP RoService Agriculture Industry   
Netherlands Agriculture   0.0003  0.0001    0.0000      0.0001   0.0001   
            Industry      0.0039  0.0010    0.0009      0.0009   0.0016   
            EHP           0.0001  0.0001    0.0001      0.0000   0.0000   
            RoService     0.0036  0.0018    0.0027      0.0005   0.0009   
OECD        Agriculture   0.0384  0.0051    0.0040      0.0031   0.0053   
            Industry      1.4955  0.1553    0.1003      0.0265   0.0941   
            EHP           0.0087  1.0388    0.0082      0.0002   0.0005   
            RoService     0.4434  0.3829    1.4204      0.0122   0.0313   
NONOECD     Agriculture   0.0090  0.0020    0.0013      1.1428   0.0880   
            Industry      0.1203  0.0320    0.0181      0.2677   1.8314   
            EHP           0.0002  0.0003    0.0001      0.0018   0.0029   
            RoService     0.0244  0.0135    0.0121      0.1824   0.3019   

                                           
                            EHP RoService  
Netherlands Agriculture  0.0000    0.0000  
            Industry     0.0008    0.0008  
            EHP          0.0000    0.0000  
            RoService    0.0007    0.0010  
OECD        Agriculture  0.0014    0.0015  
            Industry     0.0299    0.0285  
            EHP          0.0002    0.0002  
            RoService    0.0152    0.0200  
NONOECD     Agriculture  0.0214    0.0312  
            Industry     0.2891    0.2527  
            EHP          1.0124    0.0035  
            RoService    0.2618    1.3591

## Supply chain-wide effects

### demand driven/pull vs supply sided

$x* = (I-A)^{-1}y*$ 

The demand-driven (or demand pull) version of the open Input-Output model determines production as a function of final demand, given the production technology.  
- $y*$ represents a certain final demand of finished goods and services
- $x*$ represents all sectors' outputs (products/services) directly and indirectly required to fullfill the final demand for $y*$


$x = (I-A)^{-1}y$

On the contrary, in the supply-sided version, value added determines output and producers must induce sales in order to achieve a desired level of income.

- $Iy*$ > direct impact  
- $(A + A^2 + A^3 + ... + A^n)y*$ > indirect supply chain impact for n level/rounds

In [16]:
# Demand pull
y_star_index = (country, 'EHP')

y_star = pd.Series(np.array(x.index == y_star_index, dtype=int), index=x.index)
y_star

Netherlands  Agriculture    0
             Industry       0
             EHP            1
             RoService      0
OECD         Agriculture    0
             Industry       0
             EHP            0
             RoService      0
NONOECD      Agriculture    0
             Industry       0
             EHP            0
             RoService      0
dtype: int64

### Using Power series

In [17]:
# Ay*
A.dot(y_star).round(3)

Netherlands  Agriculture    0.001
             Industry       0.057
             EHP            0.034
             RoService      0.150
OECD         Agriculture    0.001
             Industry       0.018
             EHP            0.000
             RoService      0.026
NONOECD      Agriculture    0.002
             Industry       0.016
             EHP            0.000
             RoService      0.005
dtype: float64

In [18]:
# AAy*
A.dot(A.dot(y_star)).round(3)

Netherlands  Agriculture    0.002
             Industry       0.020
             EHP            0.003
             RoService      0.053
OECD         Agriculture    0.001
             Industry       0.017
             EHP            0.000
             RoService      0.020
NONOECD      Agriculture    0.002
             Industry       0.016
             EHP            0.000
             RoService      0.006
dtype: float64

In [19]:
# AAAy*
A.dot(A.dot(A.dot(y_star))).round(3)

Netherlands  Agriculture    0.001
             Industry       0.007
             EHP            0.001
             RoService      0.018
OECD         Agriculture    0.001
             Industry       0.010
             EHP            0.000
             RoService      0.013
NONOECD      Agriculture    0.001
             Industry       0.012
             EHP            0.000
             RoService      0.005
dtype: float64

In [20]:
def demand_pull_power_series(y_st, A_mat, depth):
    res = [y_st]*(depth+1)
    for i in range(1, depth+1):
        res[i] = A_mat.dot(res[i-1])

    res = pd.concat(res, axis=1)
    res.columns = ['A'*i + 'y*' for i in range(depth+1)]
    res['sum'] = res.sum(axis=1)
    return res

In [21]:
demand_pull_power_series(y_star, A, 3).round(3)

y*    Ay*   AAy*  AAAy*    sum
Netherlands Agriculture   0  0.001  0.002  0.001  0.004
            Industry      0  0.057  0.020  0.007  0.083
            EHP           1  0.034  0.003  0.001  1.038
            RoService     0  0.150  0.053  0.018  0.220
OECD        Agriculture   0  0.001  0.001  0.001  0.003
            Industry      0  0.018  0.017  0.010  0.045
            EHP           0  0.000  0.000  0.000  0.001
            RoService     0  0.026  0.020  0.013  0.059
NONOECD     Agriculture   0  0.002  0.002  0.001  0.005
            Industry      0  0.016  0.016  0.012  0.044
            EHP           0  0.000  0.000  0.000  0.000
            RoService     0  0.005  0.006  0.005  0.015

### Using Leontief inverse

Compare the power expansion with using Leontief inverse directly

In [22]:
L.dot(y_star)

Netherlands  Agriculture    0.004097
             Industry       0.086086
             EHP            1.038029
             RoService      0.228989
OECD         Agriculture    0.003359
             Industry       0.056390
             EHP            0.000952
             RoService      0.073584
NONOECD      Agriculture    0.006865
             Industry       0.062437
             EHP            0.000267
             RoService      0.023043
dtype: float64

<hr>

## Environmental accounting

### F/f: Footprint (in Tonne / euros)

$$f_i = F_i / x_i$$  

- F: (direct) emissions, resource use, etc.
- f: (direct) environmental coefficients

$$fAx + fy = fx = fLy$$  

- $fAx$: environmental pressure (EP) directly generated while producing the intermediate products
- $fy$: environmental pressure (EP) directly generated while producing the finished products
- $fx$: environmental pressure (EP) directly generated while producing all products (intermediate + finished)
- $fLy$: environmental pressure (EP) directly and inderectly generated to meet the final demand


$$f_{\text{total}} = f_{\text{direct}}L$$

- $f_{\text{total}}$ = multiplier

In [23]:
# F (Unit: Million tonne)
F_values_NL = [13.9, 92.7, 2.3, 44.7, 750.1, 7652.9, 195.1, 1991.6, 2487.6, 26232.4, 299.2, 2136.4]

F_values = F_values_NL
F = pd.Series(F_values, index=Z_columns)
F

Netherlands  Agriculture       13.9
             Industry          92.7
             EHP                2.3
             RoService         44.7
OECD         Agriculture      750.1
             Industry        7652.9
             EHP              195.1
             RoService       1991.6
NONOECD      Agriculture     2487.6
             Industry       26232.4
             EHP              299.2
             RoService       2136.4
dtype: float64

`f_direct` is a row vector of wage, job, value added, CO2 emissions, water use ... per
unit gross output. Direct environmental pressures intensity 

In [24]:
# f (Unit: tonne/Million euro)
# If measured in Million tonne/Million euro, i.e., tonne/euro, the numbers will appear too small to report. 
# coeff = F_unit / f_unit
coeff = 1_000_000

f_direct = F.divide(x)*coeff
f_direct.round(1)

Netherlands  Agriculture    415.8
             Industry       187.6
             EHP             11.4
             RoService       58.2
OECD         Agriculture    598.3
             Industry       289.3
             EHP             15.3
             RoService       52.5
NONOECD      Agriculture    582.4
             Industry       644.3
             EHP             38.4
             RoService       87.6
dtype: float64

### Using Leontief Inverse

Note: 
1) multiplier ($f_{\text{total}} = f_{\text{direct}}L$) is a row vector if we don't diagonalize f; 
2) each cell in the multiplier shows the total emissions (direct and indirect) due to one million euros of the product-country pair;

In [25]:
# multiplier (Unit: Tonne / Million euros)
f_total = f_direct.dot(L)
f_total

Netherlands  Agriculture     674.237642
             Industry        554.903051
             EHP             111.421295
             RoService       154.869998
OECD         Agriculture     803.944046
             Industry        614.432343
             EHP             126.318457
             RoService       134.152171
NONOECD      Agriculture     916.818433
             Industry       1421.309169
             EHP             324.704140
             RoService       364.659426
dtype: float64

Interpretation for NL:
- 674.2 tonnes of GHG emissions are driven by the consumption of one million euros of Dutch Agricultural products, 
- vs 803.9 and 916.8 tonnes of GHG emissions would be generated by the consumption of one million euros of Agricultural products from other OECD countries and NON-OECD countries.
- Better buy Dutch agricultural products instead of imported ones to minimize GHG emissions (including direct emissions and those along global supply chains). 

In [26]:
# for a specific country > sector
# we create y* to explore the requirements of one unit (=one million) country > sector
# products to compare the direct vs. direct + indirect in production emissions. 
y_star

Netherlands  Agriculture    0
             Industry       0
             EHP            1
             RoService      0
OECD         Agriculture    0
             Industry       0
             EHP            0
             RoService      0
NONOECD      Agriculture    0
             Industry       0
             EHP            0
             RoService      0
dtype: int64

In [27]:
# Because f is a row vector and y* is a column vector,
# the calculated results in each step is a scalar
f_total.dot(y_star)

111.42129493770204

- For EHP service products, the Dutch ones are served with fewer emission consequences (111.4 < 126.3 and 324.7). 
- The 111.4 tonnes of emissions are not limited to those emitted within the Netherlands. For example, the Dutch health services require imported products whose productions cause emissions abroad.

In [28]:
# if we don't want an aggregated value, but instead we want to bring out
# the details of where the emissions occurred by sector and country, we use
# diag(f)LY 
f_direct.mul(Id).dot(L).dot(y_star)

Netherlands  Agriculture     1.703481
             Industry       16.149826
             EHP            11.793101
             RoService      13.320615
OECD         Agriculture     2.009679
             Industry       16.313097
             EHP             0.014610
             RoService       3.862798
NONOECD      Agriculture     3.998610
             Industry       40.226326
             EHP             0.010234
             RoService       2.018919
dtype: float64

In [29]:
# and the sum of those component gives the same result
f_direct.mul(Id).dot(L).dot(y_star).sum()

111.42129493770203

### Using power series

This allow us to  compare the direct vs. direct + indirect in production each layers emissions (1, 2, 3) 

In [30]:
demand_pull_power_series(y_star, A, 3)

y*       Ay*      AAy*     AAAy*       sum
Netherlands Agriculture   0  0.000840  0.002090  0.000772  0.003701
            Industry      0  0.056697  0.019531  0.006531  0.082758
            EHP           1  0.033925  0.002981  0.000755  1.037661
            RoService     0  0.149744  0.052820  0.017591  0.220155
OECD        Agriculture   0  0.000963  0.000922  0.000667  0.002552
            Industry      0  0.017723  0.016806  0.010307  0.044836
            EHP           0  0.000203  0.000269  0.000206  0.000678
            RoService     0  0.026249  0.019848  0.012543  0.058641
NONOECD     Agriculture   0  0.001521  0.001763  0.001392  0.004677
            Industry      0  0.016009  0.016250  0.011751  0.044010
            EHP           0  0.000089  0.000053  0.000044  0.000186
            RoService     0  0.004589  0.005582  0.004603  0.014775

In [31]:
# we multiply by f
f_direct.dot(demand_pull_power_series(y_star, A, 3)).rename(index=(lambda x : 'f'+x))

fy*       11.361054
fAy*      38.772063
fAAy*     26.085995
fAAAy*    15.407403
fsum      91.626515
dtype: float64

### Production based vs Consumption based

EF represents supply chain-wide environmental pressures generated by a nation’s final demand (Footprint)

- Production-based accounting (PBA) $EF = fx + F\_Y$
  - conventional perspective
  - direct
  - territorial
- Consumption-based accounting (CBA) $EF = fLy + F\_Y$
  - direct effects are ultimately attributable to the final demand
  - direct + indirect
  - footprint

In [32]:
# GHG Footprint (Unit: Million tonne) of the region: country
# coeff = f_unit / Footprint_unit
# coeff = 1/1_000_000

Footprint = f_total.dot(Y[country])#*coeff
Footprint

241013151.9787384

In [33]:
# df_cba_pba = pd.concat([F, Footprint], axis=1).rename(
#     {0: 'PBA (territorial)', 1: 'CBA (footprint)'}, axis=1)

# df_cba_pba_percent = df_cba_pba/df_cba_pba.sum()

# print(df_cba_pba)
# print(df_cba_pba_percent)

# df_cba_pba.T.plot.bar(figsize=(10,6), stacked=True, rot=0)
# df_cba_pba_percent.T.plot.bar(figsize=(10,6), stacked=True, rot=0)

### With / Without emission from final consumers

$$EF = f_{\text{total}}y + F\_Y = f_{\text{direct}}Ly + F\_Y$$

- y doesn’t include export because the environmental pressures generated when producing exported goods and services = other countries’ footprint
- For an open economy $EF \neq F$
- For a closed economy: $f_{\text{total}}y = f_{\text{direct}}Ly = f_{\text{direct}}x = F$

In [34]:
# F_Y (Unit: million tonne)
F_Y_NL_values = [36.1,	2374.9, 3409.2]

F_Y_values = F_Y_NL_values
F_Y = pd.Series(F_Y_values, index=Y_columns, name='F_Y')
F_Y

Netherlands      36.1
OECD           2374.9
NONOECD        3409.2
Name: F_Y, dtype: float64

In [35]:
# EF (Unit: Million tonne) of the region: country
# coeff = fLY in tonnes and F_Y in million tonnes
coeff = 1/1_000_000

EF = f_total.dot(Y[country])*coeff + F_Y[country]
EF

277.1131519787384

- 277.1 Mt is larger than the footprint estimates from SRIO, with or without imports (respectively 107.2 and 182.2 Mt)
- So, we see the importance of imported goods and services in the Dutch final demand and the Dutch carbon footprint. 
- Also, the imported products must be overall more emission-intensive than the Dutch domestic products. This is because in SRIO + import, we assume the imported products have the same emission intensities as the Dutch products, which resulted in a lower GHG footprint estimate.